File: view of  Binary tree/top view.cpp
// using level order treversal with line tehnique
// time complexity: O(nlongn). Cause insertion in map is logN
// space: O(n)
class Solution
{
    public:
    //Function to return a list of nodes visible from the top view 
    //from left to right in Binary Tree.
    vector<int> topView(Node *root)
    {
        //Your code here
        vector<int> result;
        if (!root) return result;
        map<int, int> lookup;
        queue<pair<Node*, int>> q;
        Node * current = root;
        int line = 0;
        q.push({current, line});
        while (q.size() > 0){
            int levelSize = q.size();
            for (int i = 0; i < levelSize; i++){
                auto p = q.front();
                q.pop();
                current = p.first;
                line = p.second;
                if (!lookup.count(line)){
                    lookup[line] = current->data;
                }
                if (current->left){
                    q.push({current->left, line -1});
                }
                if (current->right){
                    q.push({current->right, line + 1});
                }
            }
        }
        for (auto it: lookup){
            result.push_back(it.second);
        }
        return result;
        
    }

};

File: view of  Binary tree/top view.py
# using level order treversal with line technique
# time: O(nlogn) and space: O(n)
from collections import  deque
class Solution:
    # Function to return a list of nodes visible from the top view
    # from left to right in Binary Tree.
    def topView(self, root):
        result = []
        if root is None:
            return result
        q = deque()
        lookup = {}
        current = root
        line = 0
        q.append([current, line])
        while q:
            levelSize = len(q)
            for cur in range(levelSize):
                pair = q.popleft()
                current = pair[0]
                line = pair[1]
                if line not in lookup:
                    lookup[line] = current.data
                if current.left:
                    q.append([current.left, line -1])
                if current.right:
                    q.append([current.right, line + 1])
        for key, value in sorted(lookup.items()):
            result.append(value)
        return result
